server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java

```diff
@@ -904,4 +912,86 @@ public void testDisabledByAutoExpandReplicas() {
 
         assertThat(clusterState.getRoutingNodes().shardsWithState(UNASSIGNED), empty());
     }
+
+    public void testExplanation() {
+        testExplanation(Settings.builder()
+                        .put(AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.getKey(), "zone"),
+                UnaryOperator.identity(),
+                "there are [5] copies of this shard and [2] values for attribute [zone] ([a, b] from nodes in the cluster and " +
+                        "no forced awareness) so there may be at most [3] copies of this shard allocated to nodes with each " +
+                        "value, but (including this copy) there would be [4] copies allocated to nodes with [node.attr.zone: a]");
+    }
+
+    public void testExplanationWithMissingAttribute() {
+        testExplanation(Settings.builder()
+                        .put(AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.getKey(), "zone"),
+                n -> n.add(newNode("X-0", emptyMap())),
+                "there are [5] copies of this shard and [3] values for attribute [zone] ([<missing>, a, b] from nodes in the cluster and " +
```

Comment:
On reflection I think this is a bug, we should be ignoring nulls completely, and although it didn't really matter in the past it does matter now. I opened #69374.